text/0000-extern-impl.md

```diff
@@ -0,0 +1,624 @@
+`defining::Foo` and `implementing::Foo`.
+
+There are no other constraints on the defining and implementing crates - they
+can freely define new traits, types and other items, and implementations for
+those types and traits. An implementing crate can be a defining crate for
+another crate.
+
+Regardless, the definitions within the defining crate, the implementing crate,
+and between the implementing crate and defining crate must meet the current
+coherence requirements. Also if there are multiple implementation crates in
+scope in the dependency graph, they must also be coherent.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+[Coherence](https://rust-lang.github.io/chalk/book/clauses/coherence.html) is
+currently defined in terms of "local" definitions - that is, items defined with
+in the current crate.
+
+The details of the coherence checking logic are unchanged; the only change is
+the constraints on which crates a definition and an implementation can be in.
+
+The discussion below talks about coherence in terms of the crate dependency
+relations, but the actual coherence check is at the level of individual type and
+trait implementations. The crate dependency relations set an upper bound on
+this, as any crate which does not appear within the relevent dependency graph
+need not be considered for coherence checking.
+
+## Changes to coherence checking
+
+The key problem we need to solve is how to reconcile the Rust language-level
+details (types and traits) with the build system view (dependency relationship
+between crates). Coherence is defined in terms of type/trait definitions and
+implementations, which forms its own graph. This graph is embedded in the crate
+dependency graph &mdash; that is there are no edges in the type/trait
+implementation graph which are not also present in the crate dependency graph.
+
+```mermaid
+graph TB
+  CrateC -->|dep| CrateA & CrateB
+  subgraph CrateA
+    A["Trait A"]
+    D["Type D"]
+    E["Impl A for D"] -..->|impl| A & D
+  end
+  subgraph CrateB
+    B["Type B"]
+  end
+  subgraph CrateC
+    C["Impl A for B"] -..->|impl| A & B
+  end
+```
+
+For the discussion below, crate B is visible to crate A if it falls within its
+transitive dependencies; a crate is visible to itself.
+
+```mermaid
+graph TB
+  subgraph A sees only A
+    c1["A"] --> c2["..."]
+    c3["B"]
+  end
+  subgraph A sees A&B
+    b1["A"] --> b2["..."] --> b3["B"]
+  end
+  subgraph A sees A&B
+    a1["A"] --> a2["B"]
+  end
+```
+
+The general rule is when compiling a crate, `rustc` must check the coherence of
+a set of implementations for a type if:
+- all those implementations are visible
+- no other visible crate has checked them
+
+This means the compilation of a crate must check for coherence when:
+- all the definitions and implementations are within one crate
+- if a crate A has an impl dependency on crate B, it must check coherency
+  between A and B
```

Comment:
I don't think you mean an arbitrary crate in the dependency graph for crate A here - that seems worth clarifying